### PR TITLE
DES_bs_cpt adjustments

### PR DESCRIPTION
**Pull request by solardiz**
With 3 separate commits, this fixes 2 general issues and then increases DES_bs_cpt based on benchmarks on 3 systems (AVX, AVX2, AVX-512).

### CODE REVIEW

**Comment by solardiz** [comment]
Weird, `ci/circleci: wine-64bits` failed with:

```
/usr/lib/gcc/x86_64-w64-mingw32/10.2.1/../../../../x86_64-w64-mingw32/bin/ld: /usr/x86_64-w64-mingw32/sys-root/mingw/lib/../lib/libmingwex.a(lib64_libmingwex_a-dirent.o):(.text+0x2d6): undefined reference to `__strcpy_chk'
```

**Comment by solardiz** [comment]
Turns out we need very different tuning for general-purpose CPUs with AVX-512 (try to fit in cache) vs. Knights Landing. Luckily, we can distinguish these by presence of AVX512BW:

```c
#elif defined(__AVX512BW__) && DES_BS_DEPTH == 512
#define DES_bs_cpt                      8
#else
/* Scalar, various 128-bit SIMD, Xeon Phi MIC, Xeon Phi AVX512F */
#define DES_bs_cpt                      48
#endif
```

AWS c5.24xlarge:

```
Benchmarking: descrypt, traditional crypt(3) [DES 512/512 AVX512F]... (96xOMP) DONE
Many salts:     877068K c/s real, 9136K c/s virtual
Only one salt:  87097K c/s real, 908399 c/s virtual
```

**Review by magnumripper — APPROVED** [review]

**Comment by magnumripper** [comment]
> Weird, `ci/circleci: wine-64bits` failed with:
> 
> ```
> /usr/lib/gcc/x86_64-w64-mingw32/10.2.1/../../../../x86_64-w64-mingw32/bin/ld: /usr/x86_64-w64-mingw32/sys-root/mingw/lib/../lib/libmingwex.a(lib64_libmingwex_a-dirent.o):(.text+0x2d6): undefined reference to `__strcpy_chk'
> ```

Oughta be a bug in the toolset, no? There isn't any strcpy in any of these files, neither before or after this PR.

**Comment by solardiz** [comment]
> Oughta be a bug in the toolset, no?

Right. I can force-merge this, but it'd be nasty if this issue keeps occurring after this merge.

**Comment by solardiz** [comment]
Actually even:

```
$ GOMP_CPU_AFFINITY=0-47 OMP_NUM_THREADS=48 ../run/john -test -form=descrypt
Will run 48 OpenMP threads
Benchmarking: descrypt, traditional crypt(3) [DES 512/512 AVX512F]... (48xOMP) DONE
Many salts:     924057K c/s real, 19257K c/s virtual
Only one salt:  99188K c/s real, 2066K c/s virtual
```

Limiting this to 48 threads or forks also speeds up actual cracking in a similar way, and doesn't require the affinity - it just takes a little while for the kernel to schedule the threads or processes nearly-optimally. (Our default benchmark is too quick for that.)

**Comment by solardiz** [comment]
Now this is tricky: different `cpt` is optimal depending on whether I let it use all hardware threads or only physical cores. The above was with cpt=8, but here's 16:

```
$ GOMP_CPU_AFFINITY=0-47 OMP_NUM_THREADS=48 ../run/john -test -form=descrypt
Will run 48 OpenMP threads
Benchmarking: descrypt, traditional crypt(3) [DES 512/512 AVX512F]... (48xOMP) DONE
Many salts:     939196K c/s real, 19568K c/s virtual
Only one salt:  104988K c/s real, 2184K c/s virtual
```

No easy test for this that we can perform in `john`.

**Comment by solardiz** [comment]
cpt=12 works great for 48 threads:

```
$ GOMP_CPU_AFFINITY=0-47 OMP_NUM_THREADS=48 ../run/john -test -form=descrypt
Will run 48 OpenMP threads
Benchmarking: descrypt, traditional crypt(3) [DES 512/512 AVX512F]... (48xOMP) DONE
Many salts:     944013K c/s real, 19673K c/s virtual
Only one salt:  102705K c/s real, 2146K c/s virtual
```

but hurts 96 threads, down from 877M to ~850M.

**Comment by solardiz** [comment]
cpt=10 narrows this range to 935M for 48 vs. 865M for 96.
cpt=11 gives 942M vs. 860M.
cpt=9 gives 931M (sometimes 934M) vs. 873M (sometimes 877M).

**Comment by solardiz** [comment]
With cpt=11, it's something like this (or sometimes a bit worse):

```
$ GOMP_CPU_AFFINITY=0-47 OMP_NUM_THREADS=48 ../run/john -test -form=descrypt
Will run 48 OpenMP threads
Benchmarking: descrypt, traditional crypt(3) [DES 512/512 AVX512F]... (48xOMP) DONE
Many salts:     941985K c/s real, 19649K c/s virtual
Only one salt:  102051K c/s real, 2126K c/s virtual

$ ../run/john -test -form=descrypt
Will run 96 OpenMP threads
Benchmarking: descrypt, traditional crypt(3) [DES 512/512 AVX512F]... (96xOMP) DONE
Many salts:     862912K c/s real, 9047K c/s virtual
Only one salt:  84194K c/s real, 1048K c/s virtual
```

**Comment by solardiz** [comment]
c5a.24xlarge (AMD), cpt=24 as included in this PR for AVX2 (24 is about the highest that doesn't slow down the old i7-4770K):

```
Benchmarking: descrypt, traditional crypt(3) [DES 256/256 AVX2]... (96xOMP) DONE
Many salts:     541163K c/s real, 5640K c/s virtual
Only one salt:  65731K c/s real, 686305 c/s virtual
```

48 threads is slower. Lower cpt values are slower. Higher cpt is only slightly faster (and would hurt smaller CPUs).

This is better than we had before Nov 1, which was:

```
Benchmarking: descrypt, traditional crypt(3) [DES 256/256 AVX2]... (96xOMP) DONE
Many salts:     408354K c/s real, 4262K c/s virtual
Only one salt:  64290K c/s real, 668373 c/s virtual
```

**Comment by solardiz** [comment]
I decided to use cpt=11 in AVX512BW builds. Force-pushed, and will merge.

**Comment by solardiz** [comment]
More weird CI failures here.
@claudioandre-br @magnumripper Should I just merge anyway, or do you want to experiment with CI first?

**Comment by magnumripper** [comment]
I launched "more tests" now.

On one hand we could definitely ignore that (sure as hell bogus) error, but on the other hand it will likely stick - all botchecks from now on might fail until something gets upgraded or we change something.

**Comment by ghost** [comment]
CI is ok now (a force push is needed to get the "all passed" status).

**Comment by solardiz** [comment]
I think I'll just merge and let CI run on the bleeding-jumbo branch. No need to have it run twice.

**Pull request merged**
